### PR TITLE
Initial implementation of Soul Ascetic

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: None

* **Server Mode**: Renewal

* **Description of Pull Request**: 
Taken from #7024

### CODE REVIEW

File: src/map/skill.cpp [hunk 1]
```diff
@@ -14688,6 +14787,13 @@ int skill_castend_pos2(struct block_list* src, int x, int y, uint16 skill_id, ui
 		}
 		break;
 
+	case SOA_TALISMAN_OF_BLACK_TORTOISE:
+		if (sc && (sc->getSCE(SC_T_THIRD_GOD) && !sc->getSCE(SC_T_FOURTH_GOD) && !sc->getSCE(SC_T_FIFTH_GOD))){
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:14791]:
```suggestion
		if (sc && sc->getSCE(SC_T_THIRD_GOD)){
```

According to divine-pride the status will be changed without a condition:

If the user has [Blessing of the South] buff, The buff will be changed to [Blessing of the North].

https://divine-pride.net/database/skill/5429

File: src/map/skill.cpp [hunk 2]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
+					hp += 50 * pc_checkskill( tsd, SOA_TALISMAN_MASTERY ) * sg->skill_lv;
+					hp += 50 * status_get_crt( ss ) * sg->skill_lv;
+					hp *= status_get_lv( ss ) / 100;
+
+					int sp = 0;
+
+					sp += 50 * sg->skill_lv;
+					sp += 5 * pc_checkskill( tsd, SOA_TALISMAN_MASTERY ) * sg->skill_lv;
+					sp += 5 * tstatus->crt * sg->skill_lv;
+					sp *= 100 + status_get_lv( bl );
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16801]:
```suggestion
					sp *= 100 + status_get_lv( ss );
```

According to divine-pride it should depend on the users base level and not the targets base level:

Recovers additional amount depends on level of Talisman Mastery user learned, user's base level and CRT.

https://divine-pride.net/database/skill/5422

File: src/map/skill.cpp [hunk 3]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
 			break;
+
+		case UNT_TOTEM_OF_TUTELARY:
+			++sg->val1;
+			if( bl->type == BL_PC ) {
+				if (sg->val1 % 3 == 0) {
+					int hp = 500;
+
+					hp += 500 * sg->skill_lv;
+					hp += 50 * pc_checkskill( tsd, SOA_TALISMAN_MASTERY ) * sg->skill_lv;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16792]:
```suggestion
					hp += 50 * pc_checkskill( BL_CAST( BL_PC, ss ), SOA_TALISMAN_MASTERY ) * sg->skill_lv;
```
According to divine-pride it should depend on the users level of talisman mastery and not the targets level of talisman mastery:

Recovers additional amount depends on level of Talisman Mastery user learned, user's base level and CRT.

https://divine-pride.net/database/skill/5422

File: src/map/skill.cpp [hunk 4]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
+					hp += 50 * status_get_crt( ss ) * sg->skill_lv;
+					hp *= status_get_lv( ss ) / 100;
+
+					int sp = 0;
+
+					sp += 50 * sg->skill_lv;
+					sp += 5 * pc_checkskill( tsd, SOA_TALISMAN_MASTERY ) * sg->skill_lv;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16799]:
```suggestion
					sp += 5 * pc_checkskill( BL_CAST( BL_PC, ss ), SOA_TALISMAN_MASTERY ) * sg->skill_lv;
```
According to divine-pride it should depend on the users level of talisman mastery and not the targets level of talisman mastery:

Recovers additional amount depends on level of Talisman Mastery user learned, user's base level and CRT.

https://divine-pride.net/database/skill/5422

File: db/re/skill_db.yml [hunk 5]
```diff
@@ -41846,6 +41846,725 @@ Body:
...
+          Size: 2
+        - Level: 2
+          Size: 2
+        - Level: 3
+          Size: 3
+        - Level: 4
+          Size: 3
+        - Level: 5
+          Size: 4
+      Interval: 1000
```
Comment (Lemongrass3110) [inline db/re/skill_db.yml:42090]:
```suggestion
      Interval: 500
```

According to @datawulf

File: src/map/skill.cpp [hunk 6]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
+
+		case UNT_TOTEM_OF_TUTELARY:
+			++sg->val1;
+			if( bl->type == BL_PC ) {
+				if (sg->val1 % 3 == 0) {
+					int hp = 500;
+
+					hp += 500 * sg->skill_lv;
+					hp += 50 * pc_checkskill( tsd, SOA_TALISMAN_MASTERY ) * sg->skill_lv;
+					hp += 50 * status_get_crt( ss ) * sg->skill_lv;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16793]:
```suggestion
					hp += 5 * status_get_crt( ss );
```
According to @datawulf

File: src/map/skill.cpp [hunk 7]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
+					hp *= status_get_lv( ss ) / 100;
+
+					int sp = 0;
+
+					sp += 50 * sg->skill_lv;
+					sp += 5 * pc_checkskill( BL_CAST( BL_PC, ss ), SOA_TALISMAN_MASTERY ) * sg->skill_lv;
+					sp += 5 * tstatus->crt * sg->skill_lv;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16800]:
```suggestion
					sp += 5 * tstatus->crt;
```
According to @datawulf

File: src/map/skill.cpp [hunk 8]
```diff
@@ -16675,6 +16781,32 @@ int skill_unit_onplace_timer(struct skill_unit *unit, struct block_list *bl, t_t
...
+					sp *= 100 + status_get_lv( ss );
+					sp /= 150;
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:16802]:
```suggestion
					sp *= status_get_lv( ss ) / 100;
```
According to @datawulf

File: src/map/skill.cpp [hunk 9]
```diff
@@ -8839,12 +8900,19 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
 	case PR_MAGNIFICAT:
 	case PR_GLORIA:
+	case SOA_SOUL_OF_HEAVEN_AND_EARTH:
 		if (sd == NULL || sd->status.party_id == 0 || (flag & 1)) {
 
 			// Animations don't play when outside visible range
 			if (check_distance_bl(src, bl, AREA_SIZE))
 				clif_skill_nodamage(bl, bl, skill_id, skill_lv, 1);
 
+			if( skill_id == SOA_SOUL_OF_HEAVEN_AND_EARTH ){
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:8910]:
```suggestion
			if( skill_id == SOA_SOUL_OF_HEAVEN_AND_EARTH ){
				status_percent_heal( bl, 0, 100 );

```
Missing SP healing of target.

File: src/map/skill.cpp [hunk 10]
```diff
@@ -8839,12 +8900,19 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
+				if( src != bl && tsc && tsc->getSCE(SC_TOTEM_OF_TUTELARY) ){
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:8911]:
```suggestion
				if( src != bl && sc && sc->getSCE(SC_TOTEM_OF_TUTELARY) ){
```
Using this skill while the **USER** is under the effect...

File: db/re/status.yml [hunk 11]
```diff
@@ -8883,3 +8883,79 @@ Body:
       NoBanishingBuster: true
       NoDispell: true
       NoClearance: true
+  - Status: Talisman_of_Protection
+    Icon: EFST_TALISMAN_OF_PROTECTION
+    DurationLookup: SOA_TALISMAN_OF_PROTECTION
+    Flags:
+      BlEffect: true
+      DisplayPc: true
```
Comment (Atemo) [inline db/re/status.yml:8891]: The `SC` already has these conditions
```
// If the caster is offline, dead, on another map or
// if the target is not a player or is in another party
// End the status change
```
but maybe we could also add the flag
```
	RemoveOnMapWarp       - Removed when warping to another map.
```
?

I guess the `NoSave` flag should also be defined, but that's just a guess
```
	NoSave                - Won't be saved when player logs out.
```

File: db/re/status.yml [hunk 12]
```diff
@@ -8883,3 +8883,79 @@ Body:
...
+  - Status: Talisman_of_Warrior
+    Icon: EFST_TALISMAN_OF_WARRIOR
+    DurationLookup: SOA_TALISMAN_OF_WARRIOR
+    CalcFlags:
+      Patk: true
+    Flags:
+      BlEffect: true
+      DisplayPc: true
```
Comment (Atemo) [inline db/re/status.yml:8899]: https://www.divine-pride.net/database/skill/5419
```
Can only be used on target who equips weapon.
```
How about using the flags
```
	RemoveOnUnequipWeapon - Removed when unequipping a weapon.
	RequireWeapon         - Status requires a weapon to be equipped.
```

File: db/re/status.yml [hunk 13]
```diff
@@ -8883,3 +8883,79 @@ Body:
...
+  - Status: Talisman_of_Magician
+    Icon: EFST_TALISMAN_OF_MAGICIAN
+    DurationLookup: SOA_TALISMAN_OF_MAGICIAN
+    CalcFlags:
+      Smatk: true
+    Flags:
+      BlEffect: true
+      DisplayPc: true
```
Comment (Atemo) [inline db/re/status.yml:8907]: https://www.divine-pride.net/database/skill/5420
```
Can only be used on target who equips weapon.
```
How about using the flags
```
	RemoveOnUnequipWeapon - Removed when unequipping a weapon.
	RequireWeapon         - Status requires a weapon to be equipped.
```

File: db/re/status.yml [hunk 14]
```diff
@@ -8883,3 +8883,79 @@ Body:
...
+  - Status: Talisman_of_five_Elements
+    Icon: EFST_TALISMAN_OF_FIVE_ELEMENTS
+    DurationLookup: SOA_TALISMAN_OF_FIVE_ELEMENTS
+    CalcFlags:
+      All: true
+    Flags:
+      BlEffect: true
+      DisplayPc: true
```
Comment (Atemo) [inline db/re/status.yml:8915]: https://www.divine-pride.net/database/skill/5423
```
Can only be used on target who equips weapon.
```
How about using the flags
```
	RemoveOnUnequipWeapon - Removed when unequipping a weapon.
	RequireWeapon         - Status requires a weapon to be equipped.
```

File: src/map/status.cpp [hunk 15]
```diff
@@ -4864,6 +4870,28 @@ int status_calc_pc_sub(map_session_data* sd, uint8 opt)
 		}
 		if (sc->getSCE(SC_PORK_RIB_STEW))
 			sd->dsprate -= 2;
+		if( sc->getSCE(SC_TALISMAN_OF_FIVE_ELEMENTS) ) {
+			const std::vector<e_element> elements = { ELE_FIRE, ELE_WATER, ELE_WIND, ELE_EARTH, ELE_NEUTRAL };
+			int bonus = sc->getSCE(SC_TALISMAN_OF_FIVE_ELEMENTS)->val2;
+
+			for( const auto &element : elements ){
+				sd->indexed_bonus.magic_atk_ele[element] += bonus;
```
Comment (Atemo) [inline src/map/status.cpp:4878]: https://www.divine-pride.net/database/skill/5423
```
Attaches the talisman of five elements to self or 1 party member to increases physical and magical damage against water, wind, earth, fire and neutral property enemies.
```
For me it should be
```suggestion
				sd->indexed_bonus.magic_addele[element] += bonus;
```
`magic_addele`: +x% magical damage against unit element e
while `magic_atk_ele` increases damage of e element magic of caster by x%

File: src/map/status.cpp [hunk 16]
```diff
@@ -14923,6 +14975,24 @@ TIMER_FUNC(status_change_timer){
...
+			// if the target is not a player or is in another party
+			if( ssd == nullptr || status_isdead( &ssd->bl ) || ssd->bl.m != bl->m || sd == nullptr || ssd->status.party_id != sd->status.party_id ){
+				// End the status change
+				sce->val4 = 0;
+				break;
+			}
+
+			int hp = skill_calc_heal( &ssd->bl, bl, SOA_TALISMAN_OF_PROTECTION, sce->val1, true );
+
+			status_heal( bl, hp, 0, 0, 2 );
```
Comment (Atemo) [inline src/map/status.cpp:14993]: Heal effect (not 0 HP) is also displayed on official server when player hp is hpmax.

**Comment by Masao87** [comment]
Hey there :)
Is there any news about progress on this? Currently the class bugs out on my server using the jobchange command, all previous skill points are gone from soul reaper and the new skills can't be skilled yet.
Thanks in advance for any info!

**Pull request merged**